File: src/java/br/com/lett/crawlernode/crawlers/corecontent/argentina/ArgentinaHiperlibertadCrawler.java
package br.com.lett.crawlernode.crawlers.corecontent.argentina;

import br.com.lett.crawlernode.core.models.Card;
import br.com.lett.crawlernode.core.models.Product;
import br.com.lett.crawlernode.core.models.ProductBuilder;
import br.com.lett.crawlernode.core.session.Session;
import br.com.lett.crawlernode.crawlers.extractionutils.core.VTEXNewImpl;
import br.com.lett.crawlernode.util.CrawlerUtils;
import br.com.lett.crawlernode.util.JSONUtils;
import br.com.lett.crawlernode.util.Logging;
import com.drew.lang.annotations.NotNull;
import com.google.common.collect.Sets;
import exceptions.MalformedPricingException;
import exceptions.OfferException;
import models.Offer;
import models.Offers;
import models.pricing.*;
import org.json.JSONArray;
import org.json.JSONObject;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

public class ArgentinaHiperlibertadCrawler extends VTEXNewImpl {

   public Set<String> cards = Sets.newHashSet(Card.VISA.toString(), Card.MASTERCARD.toString(),
      Card.CABAL.toString(), Card.NATIVA.toString(), Card.NARANJA.toString(), Card.AMEX.toString(), Card.CORDOBESA.toString());

   public ArgentinaHiperlibertadCrawler(@NotNull Session session) {
      super(session);
   }

   @Override
   public List<Product> extractInformation(Document doc) throws Exception {
      List<Product> products = new ArrayList<>();
      JSONObject skuInfo = scrapJsonFromHtml(doc);

      if (skuInfo != null && !skuInfo.isEmpty()) {
         Logging.printLogDebug(logger, session, "Product page identified: " + this.session.getOriginalURL());

         String internalId = skuInfo.optString("sku");
         String internalPid = skuInfo.optString("mpn");
         String name = skuInfo.optString("name");
         String description = skuInfo.optString("description");
         String primaryImage = skuInfo.optString("image");
         List<String> secondaryImages = getSecondaryImages(doc);
         String categories = skuInfo.optString("category");
         String isAvailable = JSONUtils.getValueRecursive(skuInfo, "offers.offers.0.availability", String.class, "");
         Offers offers = isAvailable != null && isAvailable.equalsIgnoreCase("http://schema.org/InStock") ? scrapOffers(skuInfo, doc) : new Offers();

         Product product = ProductBuilder.create()
            .setUrl(session.getOriginalURL())
            .setInternalId(internalId)
            .setInternalPid(internalPid)
            .setName(name)
            .setPrimaryImage(primaryImage)
            .setSecondaryImages(secondaryImages)
            .setCategories(Collections.singleton(categories))
            .setDescription(description)
            .setOffers(offers)
            .build();

         products.add(product);

      } else {
         Logging.printLogDebug(logger, session, "Not a product page " + this.session.getOriginalURL());
      }

      return products;
   }

   private List<String> getSecondaryImages(Document doc) {
      List<String> secondaryImages = new ArrayList<>();

      Elements imagesLi = doc.select("meta[property=\"og:image\"]");
      for (Element imageLi : imagesLi) {
         secondaryImages.add(imageLi.attr("content"));
      }
      if (secondaryImages.size() > 0) {
         secondaryImages.remove(0);
      }
      return secondaryImages;
   }

   private JSONObject scrapJsonFromHtml(Document doc) {
      String jsonString = CrawlerUtils.scrapScriptFromHtml(doc, "script[type=\"application/ld+json\"]:containsData(sku)");
      if (jsonString != null && !jsonString.isEmpty()) {
         JSONArray array = CrawlerUtils.stringToJsonArray(jsonString);
         if (array != null) {
            return (JSONObject) array.get(0);
         }
      }
      return null;
   }

   private Offers scrapOffers(JSONObject skuInfo, Document doc) throws MalformedPricingException, OfferException {
      Offers offers = new Offers();
      Pricing pricing = scrapPricing(skuInfo, doc);
      List<String> sales = Collections.singletonList(CrawlerUtils.calculateSales(pricing));

      offers.add(new Offer.OfferBuilder()
         .setUseSlugNameAsInternalSellerId(true)
         .setSellerFullName(this.getMainSellersNames().toString())
         .setMainPagePosition(1)
         .setIsBuybox(false)
         .setIsMainRetailer(true)
         .setPricing(pricing)
         .setSales(sales)
         .build());

      return offers;
   }

   private Pricing scrapPricing(JSONObject skuInfo, Document doc) throws MalformedPricingException {
      Double priceFrom = CrawlerUtils.scrapDoublePriceFromHtml(doc, ".vtex-product-price-1-x-currencyContainer--pdp-list-price", null, false, ',', session);
      Double spotlightPrice = JSONUtils.getValueRecursive(skuInfo, "offers.offers.0.price", Double.class, null);
      if (spotlightPrice == null) {
         Integer priceInt = JSONUtils.getValueRecursive(skuInfo, "offers.offers.0.price", Integer.class, null);
         if (priceInt != null) {
            spotlightPrice = priceInt.doubleValue();
         }
      }

      BankSlip bankSlip = CrawlerUtils.setBankSlipOffers(spotlightPrice, null);
      CreditCards creditCards = scrapCreditCards(spotlightPrice);

      return Pricing.PricingBuilder.create()
         .setSpotlightPrice(spotlightPrice)
         .setPriceFrom(priceFrom)
         .setCreditCards(creditCards)
         .setBankSlip(bankSlip)
         .build();

   }

   private CreditCards scrapCreditCards(Double price) throws MalformedPricingException {
      CreditCards creditCards = new CreditCards();
      Installments installments = new Installments();

      installments.add(Installment.InstallmentBuilder.create()
         .setInstallmentNumber(1)
         .setInstallmentPrice(price)
         .setFinalPrice(price)
         .build());

      for (String card : cards) {
         creditCards.add(CreditCard.CreditCardBuilder.create()
            .setBrand(card)
            .setInstallments(installments)
            .setIsShopCard(false)
            .build());
      }

      return creditCards;
   }
}

File: src/java/br/com/lett/crawlernode/crawlers/corecontent/argentina/ArgentinaHiperlibertadCrawler.kt
package br.com.lett.crawlernode.crawlers.corecontent.argentina

import br.com.lett.crawlernode.core.fetcher.models.Request.RequestBuilder
import br.com.lett.crawlernode.core.models.Card
import br.com.lett.crawlernode.core.models.Product
import br.com.lett.crawlernode.core.models.ProductBuilder
import br.com.lett.crawlernode.core.session.Session
import br.com.lett.crawlernode.crawlers.extractionutils.core.VTEXOldScraper
import br.com.lett.crawlernode.util.*
import models.Offer.OfferBuilder
import models.Offers
import models.RatingsReviews
import models.pricing.Pricing
import org.apache.http.impl.cookie.BasicClientCookie
import org.json.JSONArray
import org.json.JSONObject
import org.jsoup.Jsoup
import org.jsoup.nodes.DataNode
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import java.util.regex.Pattern

class ArgentinaHiperlibertadCrawler(session: Session) : VTEXOldScraper(session) {

   private val storeId: String = session.options.optString("storeId")

   override fun getHomePage(): String {
      return "https://www.hiperlibertad.com.ar/"
   }

   override fun fetch(): Any {
      if (storeId.isNotEmpty()) {
         val cookie = BasicClientCookie("VTEXSC", storeId)
         cookie.domain = "www.hiperlibertad.com.ar"
         cookie.path = "/"
         cookies.add(cookie)
      }
      val request = RequestBuilder.create().setUrl(session.originalURL).setCookies(cookies).build()
      return Jsoup.parse(dataFetcher[session, request].body)
   }

   override fun getMainSellersNames(): MutableList<String> {
      return mutableListOf("Hiper Liberdade", "LIBERTAD SA")
   }

   override fun scrapRating(internalId: String?, internalPid: String?, doc: Document?, jsonSku: JSONObject?): RatingsReviews? {
      return null
   }

   override fun extractInformation(doc: Document?): MutableList<Product> {
      val products: MutableList<Product> = ArrayList()

      val internalPid = scrapPidFromApi(doc)

      if (internalPid != null && isProductPage(doc)) {
         var productJson = crawlProductApi(internalPid, null)
         if (productJson!!.isEmpty) {
            val array = doc?.let { crawlSkuJsonArray(it) }
            productJson = array?.let { array.optJSONObject(0) }
         }
         val categories = scrapCategories(productJson)
         val description = scrapDescription(doc, productJson)
         processBeforeScrapVariations(doc, productJson, internalPid)
         if (productJson != null) {
            val items = JSONUtils.getJSONArrayValue(productJson, "items")
            if (items.length() > 0) {
               for (i in 0 until items.length()) {
                  var jsonSku = items.optJSONObject(i)
                  if (jsonSku == null) {
                     jsonSku = JSONObject()
                  }
                  val product = extractProduct(doc, internalPid, categories, description, jsonSku, productJson)
                  products.add(product)
               }
            } else {
               val name = productJson.optString("skuname")
               val internalId = productJson.optInt("sku", 0).toString()
               val images = scrapImages(doc, productJson, internalPid, internalId)
               val primaryImage: String? = if (!images.isEmpty()) images.get(0) else productJson.optString("image")
               scrapSecondaryImages(images)
               val description = CrawlerUtils.scrapStringSimpleInfo(doc, "#product-description", false)
               val offers = scrapOffers(productJson);


               val product = ProductBuilder()
                  .setUrl(session.originalURL)
                  .setInternalId(internalId)
                  .setInternalPid(internalId)
                  .setDescription(description)
                  .setName(name)
                  .setPrimaryImage(primaryImage)
                  .setSecondaryImages(images)
                  .setOffers(offers)
                  .build()

               return mutableListOf(product)
               if (product != null) {
                  products.add(product)
               }
            }
         }
      } else {
         Logging.printLogDebug(logger, session, "Not a product page " + session.originalURL)
      }

      return products
   }

   private fun crawlSkuJsonArray(document: Document): JSONArray? {
      val scriptTags = document.getElementsByTag("script")
      var skuJson: JSONObject? = null
      var skuJsonArray: JSONArray? = null
      for (tag: Element in scriptTags) {
         for (node: DataNode in tag.dataNodes()) {
            if (tag.html().trim { it <= ' ' }.startsWith("var skuJson_0 = ")) {
               skuJson = JSONObject(
                  node.wholeData.split(Pattern.quote("var skuJson_0 = ").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1]
                     + node.wholeData.split(Pattern.quote("var skuJson_0 = ").toRegex()).dropLastWhile { it.isEmpty() }
                     .toTypedArray()[1].split(Pattern.quote("}]};").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
               )
            }
         }
      }
      try {
         skuJsonArray = skuJson!!.getJSONArray("skus")
      } catch (e: Exception) {
         e.printStackTrace()
      }
      return skuJsonArray
   }

   private fun scrapOffers(json: JSONObject): Offers {
      val offers = Offers()

      val priceFromString = json.optString("listPriceFormated")
      var priceFrom: Double = MathUtils.parseDoubleWithComma(priceFromString)
      if (priceFrom == 0.00) {
         priceFrom == null
      }

      val stringPrice = json.optString("fullSellingPrice")
      val spotlightPrice: Double = MathUtils.parseDoubleWithComma(stringPrice)
      val bankSlip = spotlightPrice.toBankSlip()
      val sellerName = json.optString("seller")

      val creditCards = listOf(
         Card.MASTERCARD,
         Card.VISA,
         Card.AMEX,
         Card.NATIVA,
         Card.NARANJA,
         Card.CABAL,
         Card.CORDOBESA
      ).toCreditCards(spotlightPrice)

      offers.add(
         OfferBuilder.create()
            .setPricing(
               Pricing.PricingBuilder.create()
                  .setSpotlightPrice(spotlightPrice)
                  .setPriceFrom(priceFrom)
                  .setCreditCards(creditCards)
                  .setBankSlip(bankSlip)
                  .build()
            )
            .setSales(listOf())
            .setIsMainRetailer(true)
            .setIsBuybox(false)
            .setUseSlugNameAsInternalSellerId(true)
            .setSellerFullName(sellerName)
            .build()
      )

      return offers
   }

}
